### ISSUE TITLE
Performance of Table creation from Python lists

### ISSUE THREAD

**Issue by mcara** [astropy/astropy#10548]
This is mostly a question to understand what is causing the following performance hit in initialization of a table from Python lists:

```python
In [238]: import numpy as np
     ...: from astropy import table
     ...: time = list(np.random.randint(0, 1000000, 100000))
     ...: rate = list(np.random.randint(0, 1000000, 100000))
     ...: error = list(np.random.randint(0, 1000000, 100000))

In [239]: %timeit table.Table([time, rate, error], names=["time", "rate", "error"])
673 ms ± 14.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [240]: %timeit data = table.Table([np.array(time), np.array(rate), np.array(error)], \
     ...:                             names=["time", "rate", "error"])
17 ms ± 224 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
That was with:
```
astropy: 4.2.dev360
numpy: 1.18.0
```
With updated `numpy` to `1.19.0`, timing difference is even larger:
```
In [2]: %timeit table.Table([time, rate, error], names=["time", "rate", "error"])
861 ms ± 7.78 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit table.Table([np.array(time), np.array(rate), np.array(error)], \
   ...:                     names=["time", "rate", "error"])
16.5 ms ± 401 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

EDIT: Benchmark added in astropy/astropy-benchmarks#97

**Comment by bhayden53** [comment]
According to bisect, the timing jumps by a factor of 10 at this commit:

> 2529d00c487869c0ee5ffda23c29ff0c3d7c18bb is the first bad commit
> commit 2529d00c487869c0ee5ffda23c29ff0c3d7c18bb
> Author: Tom Aldcroft <xxxx>
> Date:   Thu Jul 25 21:09:57 2019 -0400
>     Fix bug adding a col as list of masked arrays
> :100644 100644 555575bf25117a55da0ee48ad1901228e5bf24f6 7ea37c45de3d187e988b911a1f32551c2fd7e957 M      CHANGES.rst
> :040000 040000 4ec2cae233d82295d5389b853f8bbcf14c02620a bc903a3544b772133e8f82ab7a1945aca085a6f9 M      astropy

**Comment by pllim** [comment]
Wow, you're a trooper, @bhayden53 . Thanks! That means #9048

**Comment by taldcroft** [comment]
Thanks for the investigation thus far!  This is pretty bad, I'll have a look. The root cause is `np.ma.array` (this is with 1.18.5):
```
In [1]: time = list(np.random.randint(0, 1000000, 100000))                                        

In [2]: %timeit np_data = np.ma.array(time, dtype=None)                                           
268 ms ± 11.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [3]: %timeit np_data = np.array(time, dtype=None)                                              
5.32 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
Any thoughts @mhvk on what to do here?

**Comment by saimn** [comment]
This comes from :
https://github.com/numpy/numpy/blob/18a6e3e505ee416ddfc617f3e9afdff5a031c2c2/numpy/ma/core.py#L2861-L2863
trying to get a mask from each element of the list.

```
In [17]: %timeit np_data = np.ma.array(time, dtype=None)                                                 
251 ms ± 1.48 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [18]: %timeit [np.ma.getmaskarray(m) for m in data]                                                   
188 ms ± 1.16 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [19]: %timeit [np.ma.getmaskarray(np.asanyarray(m)) for m in data]                                    
229 ms ± 2.41 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by mhvk** [comment]
That's not pretty, but I'm not sure what to do if one wants to be general and allow masked entries - one would either have to tell people who use plain lists to pre-pack with `np.array` or people who may have `np.ma.masked` or other masked arrays in their lists to pre-pack with `np.ma.MaskedArray`.

**Comment by pllim** [comment]
Hmm... Maybe add this to https://docs.astropy.org/en/latest/table/index.html#performance-tips ?

**Comment by taldcroft** [comment]
I think the use case of a plain list is far more important and common than the corner case of a list of masked arrays. The only thing I can think of for partly handling that corner case is a not-robust solution of checking for a masked array in the first element of a list and going down the masked array conversion only in that case.

**Comment by taldcroft** [comment]
Or document that case and tell people they need to explicitly create a `MaskedColumn`?  After all, the same is true of plain numpy arrays that they don't magically auto-convert.

**Comment by mcara** [comment]
I agree with https://github.com/astropy/astropy/issues/10548#issuecomment-656398073:

**Comment by mhvk** [comment]
@taldcroft - agreed that the plain list is far more common - I just thought you might fear backwards incompatibility more than you seem to (are we switching roles?? ;-), especially as the commit that introduced the slowdown was specifically to allow list of `MaskedArray`. Anyway, I'm fine with changing it, or, as you suggest, just check the first element of the list.

**Comment by pllim** [comment]
> backwards incompatibility

If you put the patch in, say, 4.2, then maybe not so much an issue. But it would be more controversial to backport to LTS. Hmm...

**Comment by taldcroft** [comment]
tl;dr - initializing `Table` with a list of things that might include masked arrays or `np.ma.masked` is a bit of a minefield with inconsistent behaviors.

I did some digging in an effort to fix this. It turns out to be a little bit trickier than I had imagined. The changes in #9048 (basically converting a list to array via MaskedArray) enabled a new behavior which previously didn't work, namely initializing from a list that contains `np.ma.masked`.

One commit before #9048:
```
(astropy-temp) ➜  astropy-temp git:(5a791cffd) ✗ ipython
Python 3.7.7 (default, May  6 2020, 04:59:01) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: astro                                                                                                                                             
astropy=4.0.dev25421

In [2]: Table([[1.0, np.ma.masked], ['a', np.ma.masked]])                                                                                                 
/Users/aldcroft/git/astropy-temp/astropy/table/column.py:221: UserWarning: Warning: converting a masked element to nan.
  self_data = np.array(data, dtype=dtype, copy=copy)
Out[2]: 
<Table length=2>
  col0   col1
float64 str32
------- -----
    1.0     a
    nan   0.0
```
After #9048
```
(astropy-temp) ➜  astropy-temp git:(5a791cffd) ✗ git checkout 2529d00c487869c0ee5f
M	astropy/table/tests/test_masked.py
Previous HEAD position was 5a791cffd Merge pull request #9035 from adrn/coordinates/organize-api-docs
HEAD is now at 2529d00c4 Fix bug adding a col as list of masked arrays
(astropy-temp) ➜  astropy-temp git:(2529d00c4) ✗ ipython                          
Python 3.7.7 (default, May  6 2020, 04:59:01) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.16.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: astro                                                                                                                                             
astropy=4.0.dev25422

In [2]: Table([[1.0, np.ma.masked], ['a', np.ma.masked]])                                                                                                 
/Users/aldcroft/miniconda3/envs/astropy-temp/lib/python3.7/site-packages/numpy/ma/core.py:2788: UserWarning: Warning: converting a masked element to nan.
  order=order, subok=True, ndmin=ndmin)
Out[2]: 
<Table length=2>
  col0   col1
float64 str32
------- -----
    1.0     a
     --    --
```

This behavior was not tested at that point, but became part of regression tests later on with #9651, specifically:
```
    def test_initialization_with_all_columns(self):
        t1 = Table([self.a, self.b, self.c, self.d, self.ca, self.sc])
        assert t1.colnames == ['a', 'b', 'c', 'd', 'ca', 'sc']
        # Check we get the same result by passing in as list of dict.
        # (Regression test for error uncovered by scintillometry package.)
        lofd = [{k: row[k] for k in t1.colnames} for row in t1]
        t2 = Table(lofd)
        for k in t1.colnames:
            assert np.all(t1[k] == t2[k]) in (True, np.ma.masked)
            assert np.all(getattr(t1[k], 'mask', False)
                          == getattr(t2[k], 'mask', False))
```

However, "working" is a little bit of an overstatement. The underlying conversion with `MaskedArray` fails for `int` type:
```
In [11]: np.ma.MaskedArray([1, np.ma.masked], dtype=int)                                                                         
MaskError: Cannot convert masked element to a Python int.

# Automatically coerces to float if no dtype is provided
In [12]: np.ma.MaskedArray([1, np.ma.masked])                                                                                    
/Users/aldcroft/miniconda3/envs/astropy/lib/python3.7/site-packages/numpy/ma/core.py:2795: UserWarning: Warning: converting a masked element to nan.
  order=order, subok=True, ndmin=ndmin)
Out[12]: 
masked_array(data=[1.0, --],
             mask=[False,  True],
       fill_value=1e+20)
```

In fact if you add a test `assert t1[k].dtype == t2[k].dtype` to the above test it fails. For instance the `dtype` of `t2['b']` is object not int64.

**Comment by taldcroft** [comment]
FWIW, pandas has a similar behavior:
```
In [11]: Series([1,np.ma.masked])                                                                                                
Out[11]: 
0     1
1    --
dtype: object
```

**Comment by pllim** [comment]
p.s. I don't see the new benchmark result being reported yet at http://www.astropy.org/astropy-benchmarks/
